src/Catalyst.Core.Lib.Tests/UnitTests/IO/Messaging/Correlation/MessageCorrelationManagerTests.cs

```diff
@@ -0,0 +1,232 @@
+
+        protected void FireEvictionCallBackByCorrelationId(ByteString correlationId)
+        {
+            var pendingRequest = PendingRequests.Single(p => p.Content.CorrelationId == correlationId);
+            CacheEntriesByRequest[correlationId].PostEvictionCallbacks[0].EvictionCallback
+               .Invoke(null, pendingRequest, EvictionReason.Expired, null);
+        }
+
+#pragma warning disable 693
+        protected void TryMatchResponseAsync_Should_Match_Existing_Records_With_Matching_Correlation_Id<T>()
+#pragma warning restore 693
+            where T : IMessage, new()
+        {
+            var responseMatchingIndex1 = new T().ToProtocolMessage(
+                PeerIds[1],
+                PendingRequests[1].Content.CorrelationId.ToCorrelationId());
+
+            var request = CorrelationManager.TryMatchResponse(responseMatchingIndex1);
+            request.Should().BeTrue();
+        }
+
+#pragma warning disable 693
+        protected void TryMatchResponseAsync_Should_Not_Match_Existing_Records_With_Non_Matching_Correlation_Id<T>()
+#pragma warning restore 693
+            where T : IMessage, new()
+        {
+            var responseMatchingNothing =
+                new T().ToProtocolMessage(PeerIds[1], CorrelationId.GenerateCorrelationId());
+
+            var request = CorrelationManager.TryMatchResponse(responseMatchingNothing);
+            request.Should().BeFalse();
+        }
+
+        [Fact]
+        public void UncorrelatedMessage_Should_Not_Propagate_To_Next_Pipeline()
+        {
+            var correlationManager = Substitute.For<IMessageCorrelationManager>();
+            correlationManager.TryMatchResponse(Arg.Any<ProtocolMessage>()).Returns(false);
+
+            var correlationHandler = new CorrelationHandler<IMessageCorrelationManager>(correlationManager);
+            var channelHandlerContext = Substitute.For<IChannelHandlerContext>();
+            var nonCorrelatedMessage =
+                new PingResponse().ToProtocolMessage(PeerIds[0], CorrelationId.GenerateCorrelationId());
+            correlationHandler.ChannelRead(channelHandlerContext, nonCorrelatedMessage);
+
+            channelHandlerContext.DidNotReceive().FireChannelRead(nonCorrelatedMessage);
+        }
+
+        [Fact]
+        public void TryMatchResponseAsync_Should_Not_Match_On_Wrong_Response_Type()
+        {
+            var matchingRequest = PendingRequests[2].Content;
+            var matchingRequestWithWrongType =
+                new PeerNeighborsResponse().ToProtocolMessage(matchingRequest.PeerId,
+                    matchingRequest.CorrelationId.ToCorrelationId());
+
+            new Action(() => CorrelationManager.TryMatchResponse(matchingRequestWithWrongType))
+               .Should().Throw<InvalidDataException>()
+               .And.Message.Should().Contain(PeerNeighborsResponse.Descriptor.ShortenedFullName());
+        }
+
+        protected virtual void Dispose(bool disposing)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Either implement 'IDisposable.Dispose', or totally rename this method to prevent confusion.](https://app.codacy.com/app/catalyst-network/Catalyst.Node/pullRequest?prid=4318201)